### ISSUE TITLE
Pen / Stylus Eraser support (on button hold)

### ISSUE THREAD

**Issue by TomasKliner** [excalidraw/excalidraw#5166]
It would be amazing to have support to **activate eraser, when stylus / pen eraser button is pressed** and disable/switch back to previous tool, when button is released.

_Currently it is the only feature holding me back from switching from OneNote, since its dealbreaker for touchscreen/stylus._

**Comment by dwelle** [comment]
Ok, so I dug around a bit and this might be doable, but heavily depends on the tablet/styles/OS. What tabled and OS are you using?

For example, on Wacom, the eraser on the other end of the stylus:

![image](https://user-images.githubusercontent.com/5153846/167488685-2f6278cd-678f-43a4-829b-3d72898f1078.png)

emits `event.button === 5` which we can work with. 

But, on Windows it works only if the `Use Windows Ink` is enabled:

![image](https://user-images.githubusercontent.com/5153846/167492303-65aa282d-6504-4107-a95b-11653a35518a.png)

Otherwise it doesn't emit any events.

The stylus side buttons are by default mapped to generic buttons that we can't/don't want to use as an eraser (right click for opening contextmenu, and a double click). 

![image](https://user-images.githubusercontent.com/5153846/167488859-ba72d37c-d601-4e73-9771-a2a941579fe7.png)

Trying to remap them to something useful doesn't work on Windows _unless_ you disable the Windows Ink feature, which in turn makes the Eraser button not work :).

May work differently on MacOS.

**Comment by TomasKliner** [comment]
Hello, i mainly use Windows with lenovo Active Pen. I would say it have wacoom like settings,  with 2 side buttons, lower one used for eraser. Same goes for my main machine with One by Wacoom.
It would be amazing if you could somehow detect the 5th button.

But, when i did some digging to figure out if it emmits any event.button/s, I couldnt get it to work. Not sure if browsers doesnt support it, or if its just my type stylus (maybe Windows Ink)? Tried both on Linux and WIndows.

**Comment by TomasKliner** [comment]
Alternativly it may be worth to make button shortcut for eraser, that would use eraser while pressed down and get back to previous tool/pen on release. Since most of tablets have settings to change/remap buttons based on current application.

**Comment by dwelle** [comment]
You can use this to check whether your pen buttons emit `pointerdown` events and what `button` id it does. https://vw6l9k.csb.app/

**Comment by TomasKliner** [comment]
Yeah still the same result without emitting the Erase button, even tried setting the button specifficaly in settings insted of Eraser to Emmiting button 5 without success.
Im going to check on my Wacoom tablet if it behaves the same once i get to it.

Not sure if its not supported in JS implementation, since OneNote detect it somehow without issue?

**Comment by TomasKliner** [comment]
Well its sad, that in 2022 there is no better support of Pen/Eraser for Developers, since apps like yours it would greatly benefit from it, anyway appreciate your work !

**Comment by TomasKliner** [comment]
> Alternativly  it may be worth to make button shortcut for eraser, that would use eraser while pressed down and get back to previous tool/pen on release. Since most of tablets have settings to change/remap buttons based on current application.
 
As for somewhat of a **workaround suggestion**:
Illustration how it could work:
<img src="https://user-images.githubusercontent.com/50572595/167631132-8adea223-6896-4c67-a6f9-0e4d72ba1b7c.png" height="200" />

I feel like two different styles of erase button shortcuts would be good even without the Pen/remapped buttons. for better UX (less friction, since the switch would be much faster and without thinking).

**Comment by TomasKliner** [comment]
Thats true, but it would probably also require some fine tunning to get the delay right, since it would bring some error. 
**Personal opinion:**
1 sec seems way to high for quickly erasing one line with mistake and continuing to draw.
If you decide to go this way this website might be helpfull to test the right delay https://blog.seethis.link/scan-rate-estimator/.
For me personally the average button press is around 90 ms And somewhere around 300-400 it starts to feel long. So 500 could be a nice.

**Comment by dwelle** [comment]
> Thats true, but it would probably also require some fine tunning to get the delay right, since it would bring some error.

Yeah. It'd not be the most ideal solution because when you're in the flow you might be doing a very short bursts of erasing, possibly even below 500ms (and yes, 1 sec is too much now that I'm trying it out!).

**Comment by TomasKliner** [comment]
Also i tried "One by Wacoom" tablet and its able to emmit button 5. So it may be more device specific bug on my other device. Will do more tinkering to try and get it to work.

**Comment by dtinth** [comment]
I created a workaround that works with Micrsoft Surface Pen and Microsoft Edge. It might work with a Wacom tablet because it also emits button 5th (which is the same button as a Surface Pen).

- https://github.com/excalidraw/excalidraw/issues/5281#issue-1261476402

**Comment by HiiGHoVuTi** [comment]
Was this added? My Samsung S Pen does not erase when I press the button. However the button is detected, because instead of enabling the eraser it acts as a right-click button.
Is there a way to configure it?

**Comment by hpr1999** [comment]
> Was this added? My Samsung S Pen does not erase when I press the button. However the button is detected, because instead of enabling the eraser it acts as a right-click button.
> Is there a way to configure it? 

I'd like to second the question :) I'd love to use Excalidraw for my workflow and erasing makes it much more fluid.

**Comment by wokawoka** [comment]
> > Was this added? My Samsung S Pen does not erase when I press the button. However the button is detected, because instead of enabling the eraser it acts as a right-click button.
> > Is there a way to configure it?
> 
> I'd like to second the question :) I'd love to use Excalidraw for my workflow and erasing makes it much more fluid.

+1

**Comment by LicanMartinez** [comment]
> Unfortunately, Excalidraw inside Obsidian (Android Samsung Galaxy Tab S6) doesn't have the option of erasing.
> 
> ![image](https://user-images.githubusercontent.com/4195550/200654000-e26f8ef1-5e22-4081-aa9d-53bec203c8a3.png)

Android version for Samsung S21 Ultra shows it falling off the edge, and it's not a side-scrollable menu. Luckily the half-showing is enough to reach to it with the S pen. 
![Screenshot_20221222_221345](https://user-images.githubusercontent.com/66184913/209251664-b9f1577e-6535-46a2-9a60-8a2391a78f13.jpg)

Also it would be very useful to have erase access when pressing Spen button.

**Comment by githubbapoopa** [comment]
ThinkPad X1 Yoga
Integrated pen
Wacom Components Driver version: 7.7-61  (https://us.wacom.com/en/feeldriver/)
I use the slightly newer Lenovo Pen Settings app https://apps.microsoft.com/store/detail/lenovo-pen-settings/9MVLWT5DMSKR?hl=en-us&gl=US (which is only a rebranded Wacom Pen Settings app https://www.microsoft.com/en-us/p/wacom-pen/9p32cf64v70c?activetab=pivot:overviewtab > they work the same)


Upper button: click
Lower button: eraser

![sx_Lenovo_Pen_Settings_2023 01 02_20 26 13 304](https://user-images.githubusercontent.com/82314033/210272790-90394b09-2402-4733-948e-4e32c96ec5f1.png)

This gives me (on this side https://vw6l9k.csb.app/):
Upper button: 5
Lower button: 2

But *only* on Firefox 102.6.0esr

How to get the Eraser (+ other events) working from that point onwards in obsidian?

**Comment by Reno0vacio** [comment]
Is it not possible to create a script that captures the incoming bluetooth signal transmitted by the eraser function from the pen?

**Comment by etienn8** [comment]
> > > Was this added? My Samsung S Pen does not erase when I press the button. However the button is detected, because instead of enabling the eraser it acts as a right-click button.
> > > Is there a way to configure it?
> > 
> > 
> > I'd like to second the question :) I'd love to use Excalidraw for my workflow and erasing makes it much more fluid.
> 
> +1

+1 
For me, that feature will be a game changer. I'm currently using OneNote and I don't transfer to Obsidian because the button of the s-pen doesn't trigger  the eraser in Excalidraw. I'm convinced that I will switch to Obsidian if you support it :)

**Comment by adman44532** [comment]
+1
Agreed, I use this feature in OneNote all the time and would love to see it implemented for Excalidraw especially for the obsidian plugin.

**Comment by githubbapoopa** [comment]
> Is it not possible to create a script that captures the incoming bluetooth signal transmitted by the eraser function from the pen?

I am not using Bluetooth in my workflow at all. It never worked reliably with any Wacom drivers (+Bamboo Ink Plus).

My question is soley for the "normal" AES2 digitizer of my screen.

But that shouldn't matter in the end as the buttons all trigger events.

Nobody?

**Comment by iMagUdspEllr** [comment]
When I hold down the s pen button, it erases. But, it doesn't switch back to drawing when I release the button. Is there something that I'm doing wrong?

**Comment by ctwhome** [comment]
Will this feature be planned? It would be nice to hear from the core team

**Comment by njalooo** [comment]
+1 , even a simple press and release Key for the Eraser, which can be set in Obsidian's Hotkey settings would be great.

**Comment by Le0X8** [comment]
That's currently the only one missing feature that keeps me using Samsung Notes.

Would be great if that will be actually a thing!

**Comment by Squeakrats** [comment]
I wanted this as well so I implemented it in #7348. 
Works as expected and I can erase everything using the eraser side of my stylus.

**Comment by ctwhome** [comment]
This will never be implemented for S-pen as Samsung doesn't emit button events for the s-pen on websites. For that it needs to be use a native android module or app that captures those events and send it to excalidraw. I don't know how the s-pen sdk is integrated on windows machines like the Samsung galaxy book.

**Comment by Squeakrats** [comment]
As Excalidraw is primarily a web app, it should support all “web standard conforming” input devices. If vendors/manufacturers are not writing good drivers or for whatever other reason, it’s probably just “out of scope”. You can test that “conforming” devices have functional erasers in other drawing apps like PowerPoint online.

**Comment by arthurrump** [comment]
Looks like this is fixed for well-behaving systems that report the eraser functionality correctly in #7496 by @dwelle.

My One by Wacom on Linux in Firefox is not well-behaving, and I suppose there are quite a few other weird systems that aren't either, so I think the workaround suggested [before in this thread](https://github.com/excalidraw/excalidraw/issues/5166#issuecomment-1122528652) would still be nice to have:

> As for somewhat of a **workaround suggestion**: Illustration how it could work: <img alt="" height="200" src="https://user-images.githubusercontent.com/50572595/167631132-8adea223-6896-4c67-a6f9-0e4d72ba1b7c.png">
> 
> I feel like two different styles of erase button shortcuts would be good even without the Pen/remapped buttons. for better UX (less friction, since the switch would be much faster and without thinking).

This is similar to how the spacebar temporarily switches to the Hand-tool, so it's not entirely out of place. Ideally this would be configurable to any input, so that no remapping is needed on the tablet-side and its functionality does not change in other applications, but maybe handling arbitrary configured inputs is a little out of scope for Excalidraw.

In the meantime, I've found a workaround on Linux using [Input Remapper](https://github.com/sezanzeb/input-remapper/). Mapping the stylus button to `key(e).hold().key(e)` will switch to the eraser when pressing the chosen button and revert back to the previous tool on release.

![image](https://github.com/excalidraw/excalidraw/assets/10365257/e7db45cf-026f-4450-baab-ab3cac3c72b5)

With an erase-while-hold shortcut directly in Excalidraw, this would be simpler and more likely to work with simpler tools and in other situations as well.

**Comment by ahmadshiddiq** [comment]
is there any updates for samsung tablet?

**Comment by buylife** [comment]
How about Samsung Fold and S-Pen? Thank you

**Comment by AlphaCraft9658** [comment]
Any updates? I would like for use the button of the S-Pen, or maybe even gestures, if possible.

**Comment by dacevedo12** [comment]
Lenovo Precision Pen 2, anyone?

**Comment by luanlouzada** [comment]
s-pen pls, up

**Comment by magmabits** [comment]
S-Pen; +1
